### PR TITLE
ds-lite: Add support for IPIP6(RFC2473) tunnel

### PR DESCRIPTION
**Pull request by missing233**
Add Generic Packet Tunneling in IPv6 Specification (RFC 2473) support.
(For NTT Broadband V6Plus/transix/Xpass Fixed IP)

[depends on other PRs: openwrt/luci](https://github.com/openwrt/luci/pull/6875)

### CODE REVIEW

**Comment by missing233** [comment]
@981213 
I have reopened a PR and would appreciate it if you could check it.
Additionally, I have reconsidered the issue of integrating with the ds-lite or ipip packages. If integrated with ds-lite, since the ds-lite tunnel is a specialization of this ipip6 tunnel, it seems that ds-lite should be renamed to an interface name prefix like ipip6-ds-xxx. However, this might lead to conflicts because past setups using the dslite-xxx interface name could potentially result in errors if updated to ipip6. Therefore, I have decided to keep the ipip6 package separate from ds-lite. Do you have any better suggestions?

**Comment by 981213** [comment]
For compatibility reason the existing ds-lite protocol name and interface name shouldn't be changed. You can extract the shared code and register two separated protocols. See package/network/config/gre for an example.
With so much copied code I think it's better to just put them together instead of duplicating the script.

**Comment by missing233** [comment]
> For compatibility reason the existing ds-lite protocol name and interface name shouldn't be changed. You can extract the shared code and register two separated protocols. See package/network/config/gre for an example.
> With so much copied code I think it's better to just put them together instead of duplicating the script.

Currently, ds-lite and ipip6 are two separate packages. Do you think it would be better to merge them into a single package?

**Comment by 981213** [comment]
> Do you think it would be better to merge them into a single package?

I think it's fine for them to live in a single package, considering the reduced code duplication.

**Comment by missing233** [comment]
> > Do you think it would be better to merge them into a single package?
> 
> I think it's fine for them to live in a single package, considering the reduced code duplication.

Sorry for the late reply. I have merged the backends of DS-Lite and IPIP6 into a single backend. 
I did not remove the DS-Lite package but replaced dslite.sh with ipip6.sh from the IPIP6 package.

**Comment by systemcrash** [comment]
How's this coming along?

**Comment by missing233** [comment]
> How's this coming along?

It's essentially complete, but it hasn't been reviewed or merged by anyone yet.

File: package/network/ipv6/ds-lite/files/ipip6.sh
Comment (981213) [inline package/network/ipv6/ds-lite/files/ipip6.sh:135]: Extra empty line

File: package/network/ipv6/ds-lite/Makefile
Comment (981213) [inline package/network/ipv6/ds-lite/Makefile:39]: This, and the renaming of dslite.sh to ipip6.sh, can go into a separated commit.
It's incomplete to only rename this script and keep all the other ds-lite names around.
You could either leave the file name unchanged, or rename the entire package to ipip6 and add a
```
PROVIDES:=ds-lite
```
line using a separated commit.

**Review by 981213 — CHANGES_REQUESTED** [review]
Hi!
You don't need to add another ipip6 package for a duplicated script, since the whole point of merging scripts is to prevent having two copies of duplicated code.
Three more comments below:

**Comment by missing233** [comment]
> Hi! You don't need to add another ipip6 package for a duplicated script, since the whole point of merging scripts is to prevent having two copies of duplicated code. Three more comments below:

Thank you for the review. I've made the requested changes.

**Comment by 981213** [comment]
> Thank you for the review. I've made the requested changes.

You did the opposite of what I said, which made the history of ds-lite package harder to track...
You should amend your commit and force push the branch. There should be exactly two commits in this PR.

**Review by 981213 — CHANGES_REQUESTED** [review]
Hi! The first commit ("ipip6: add new package for IPIP6(RFC2473) support") should be dropped. It adds files which are immediately deleted in the second commit and it's unnecessary to have such a change in the commit history.
The second commit should be reworded to something like:
ds-lite: add support for IPIP6(RFC2473)
and you should also bump the package version of ds-lite in Makefile.

One more comment below:

**Comment by missing233** [comment]
> Hi! The first commit ("ipip6: add new package for IPIP6(RFC2473) support") should be dropped. It adds files which are immediately deleted in the second commit and it's unnecessary to have such a change in the commit history. The second commit should be reworded to something like: ds-lite: add support for IPIP6(RFC2473) and you should also bump the package version of ds-lite in Makefile.
> 
> One more comment below:

Thanks! I've implemented the suggested changes.

File: package/network/ipv6/ds-lite/files/dslite.sh
Comment (systemcrash) [reply package/network/ipv6/ds-lite/files/dslite.sh]: valid, fine, acceptable.

**Comment by systemcrash** [comment]
> Hi! The first commit ("ipip6: add new package for IPIP6(RFC2473) support") should be dropped. It adds files which are immediately deleted in the second commit and it's unnecessary to have such a change in the commit history. The second commit should be reworded to something like: ds-lite: add support for IPIP6(RFC2473) and you should also bump the package version of ds-lite in Makefile.

Is the Makefile supposed to have ds-lite AND ipip6 or just one or the other?

**Comment by missing233** [comment]
> Hi! The first commit ("ipip6: add new package for IPIP6(RFC2473) support") should be dropped. It adds files which are immediately deleted in the second commit and it's unnecessary to have such a change in the commit history. The second commit should be reworded to something like: ds-lite: add support for IPIP6(RFC2473) and you should also bump the package version of ds-lite in Makefile.
> 
> One more comment below:

Could you kindly let me know when you might have time to review my PR?

File: package/network/ipv6/ds-lite/files/dslite.sh
Comment (981213) [inline package/network/ipv6/ds-lite/files/dslite.sh:3]: I remember there were discussions about how this OpenWrt.org copyright isn't valid and we all stopped updating them at one point. You can leave this as-is.

File: package/network/ipv6/ds-lite/files/dslite.sh
Comment (981213) [inline package/network/ipv6/ds-lite/files/dslite.sh:4]: You don't need an extra line here.

File: package/network/ipv6/ds-lite/files/dslite.sh
Comment (981213) [inline package/network/ipv6/ds-lite/files/dslite.sh:90]: The indentation of this block looks weird. Could you replace 4 spaces with tabs instead?

File: package/network/ipv6/ds-lite/Makefile
Comment (981213) [inline package/network/ipv6/ds-lite/Makefile:20]: As @systemcrash pointed out, this is a bit ambiguous.
You can use 'and' instead of '/' here.
Same for the description a few lines below.

**Review by 981213 — COMMENTED** [review]
I'm sorry for more comments again, but I noticed some cosmetic issues below:

**Comment by missing233** [comment]
> I'm sorry for more comments again, but I noticed some cosmetic issues below:

Okay, it's all done.

**Review by 981213 — APPROVED** [review]

**Comment by systemcrash** [comment]
I'm just checking with @aparcar about maintainers having multiple users, but I think this can go in as is. This PR replaces @sbyx.

**Review by systemcrash — APPROVED** [review]

**Comment by missing233** [comment]
> I'm just checking with @aparcar about maintainers having multiple users, but I think this can go in as is. This PR replaces @sbyx.

Thanks :)
As this is my first PR to openwrt, I'm still learning the ropes. If there are any issues, please let me know!

**Comment by systemcrash** [comment]
What's the status of this PR? Is it good to go? Multiple maintainers is OK, but they don't all show up in packet search.

**Comment by missing233** [comment]
> What's the status of this PR? Is it good to go? Multiple maintainers is OK, but they don't all show up in packet search.

It has been completed, but no one has come to merge it.

**Comment by missing233** [comment]
> What's the status of this PR? Is it good to go? Multiple maintainers is OK, but they don't all show up in packet search.

It has been completed, but no one has come to merge it. I'm not sure what I should do.

File: package/network/ipv6/ds-lite/Makefile
Comment (981213) [inline package/network/ipv6/ds-lite/Makefile:21]: maintainer takeover should happen in a separated commit.

File: package/network/ipv6/ds-lite/files/dslite.sh
Comment (981213) [inline package/network/ipv6/ds-lite/files/dslite.sh:33]: That's a subshell. I don't think it should be dropped.

File: package/network/ipv6/ds-lite/files/dslite.sh
Comment (981213) [inline package/network/ipv6/ds-lite/files/dslite.sh:45]: This is unrelated to adding ipip6 support.

File: package/network/ipv6/ds-lite/files/dslite.sh
Comment (981213) [inline package/network/ipv6/ds-lite/files/dslite.sh:75]: Also unrelated.

**Review by 981213 — COMMENTED** [review]
Hello!
There are a bunch of changes unrelated to adding ipip6 support which I think should happen in separated commits.
As it's been quite a while for this PR to be merged I dropped them myself and merged it into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git;a=summary
Thanks!

**Comment by missing233** [comment]
> Hello!
> There are a bunch of changes unrelated to adding ipip6 support which I think should happen in separated commits.
> As it's been quite a while for this PR to be merged I dropped them myself and merged it into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git;a=summary
> Thanks!

Hi, I am a bit confused regarding the mention of unrelated changes in my PR. To clarify, the PR I submitted was focused solely on adding IPIP6 support, and I did not include any other changes or features. Could you please elaborate on the unrelated changes you mentioned?

do I need to reopen a new PR?

**Comment by 981213** [comment]
> Could you please elaborate on the unrelated changes you mentioned?

I've put what I decided to drop in the review comments above.
A commit should only do what the commit message says it does. In this case, a commit titled "adding support for IPIP6" shouldn't replace the maintainer, replace unrelated code segments, or change whitespaces.
They should go into separated commits, each with commit message explaining why it's necessary to do so.

> do I need to reopen a new PR?

Or just respond to the comments above? I can still update the commit as I haven't pushed it into the main repo.

**Comment by missing233** [comment]
> > Could you please elaborate on the unrelated changes you mentioned?
> 
> I've put what I decided to drop in the review comments above. A commit should only do what the commit message says it does. In this case, a commit titled "adding support for IPIP6" shouldn't replace the maintainer, replace unrelated code segments, or change whitespaces. They should go into separated commits, each with commit message explaining why it's necessary to do so.
> 
> > do I need to reopen a new PR?
> 
> Or just respond to the comments above? I can still update the commit as I haven't pushed it into the main repo.

I have resolved the issues you mentioned. However, since you have closed this PR, it seems that the changes won't be synchronized here.

File: package/network/ipv6/ds-lite/Makefile
Comment (missing233) [reply package/network/ipv6/ds-lite/Makefile:21]: It has now been reverted back to the previous maintainer.

**Comment by 981213** [comment]
> I have resolved the issues you mentioned. However, since you have closed this PR, it seems that the changes won't be synchronized here.

The version I applied is identical to your current version, so I think there's nothing to be done here :D
I'll push it to the main branch this weekend.

**Comment by missing233** [comment]
> > I have resolved the issues you mentioned. However, since you have closed this PR, it seems that the changes won't be synchronized here.
> 
> The version I applied is identical to your current version, so I think there's nothing to be done here :D I'll push it to the main branch this weekend.

Thanks. After you merge it into the main branch, do I need to change this [PR](https://github.com/openwrt/luci/pull/6875) to "Ready for review"?

**Comment by 981213** [comment]
> do I need to change this [PR](https://github.com/openwrt/luci/pull/6875) to "Ready for review"?

Yes.

**Pull request closed without merging**
